### PR TITLE
mgr/dashboard: Update datatable only when necessary

### PR DESCRIPTION
**Pull request by votdev**
Fixes: https://tracker.ceph.com/issues/47359

Signed-off-by: Volker Theile <vtheile@suse.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts [hunk 1]
```diff
@@ -584,8 +586,14 @@ export class TableComponent implements AfterContentChecked, OnInit, OnChanges, O
     return _.isEmpty(css) ? undefined : css;
   }
 
-  ngOnChanges() {
-    this.useData();
+  ngOnChanges(changes: SimpleChanges) {
+    _.forIn(changes, (_change: SimpleChange, propName: string) => {
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts:590]: I find the native `for-of` looping simpler:
```suggestion
    for (const propName in changes) {
```
Or even [simpler](https://dev.to/juanjoms/comment/10511):
```ts
ngOnChanges(changes: SimpleChanges) {
  if (changes.data && changes.data.currentValue) {
    this.userData();
  }
}
```

**Review by epuertat — APPROVED** [review]
Looks good. Thanks for taking care of this. Just commented on the syntax used.

**Comment by votdev** [comment]
jenkins retest this please

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts [hunk 2]
```diff
@@ -584,8 +585,10 @@ export class TableComponent implements AfterContentChecked, OnInit, OnChanges, O
     return _.isEmpty(css) ? undefined : css;
   }
 
-  ngOnChanges() {
-    this.useData();
+  ngOnChanges(changes: SimpleChanges) {
+    if (changes.data && changes.data.currentValue) {
```
Comment (tspmelo) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts:589]: Not required suggestion:
```suggestion
    if (changes.data?.currentValue) {
```

**Review by tspmelo — APPROVED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts (hunk 2 above)
Comment (bk201) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts:589]: I really like the syntactic sugar, but it makes backport not straightforward. :(

**Comment by LenzGr** [comment]
jenkins test dashboard

**Comment by LenzGr** [comment]
jenkins test api

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts (hunk 2 above)
Comment (votdev) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.ts:589]: I think we should keep the old style to make backporting easier.

**Comment by votdev** [comment]
jenkins test dashboard

**Comment by votdev** [comment]
jenkins test dashboard

**Review by callithea — APPROVED** [review]
lgtm

**Comment by callithea** [comment]
jenkins test dashboard

**Pull request merged**